lib/ansible/module_utils/network/nxos/config/l2_interfaces/l2_interfaces.py

```diff
@@ -255,7 +255,10 @@ def add_commands(self, d):
         if 'vlan' in d:
             commands.append(cmd + 'access vlan ' + str(d['vlan']))
         if 'allowed_vlans' in d:
-            commands.append(cmd + 'trunk allowed vlan ' + str(d['allowed_vlans']))
+            if vlan_exists:
```

Comment:
I just ran a test and discovered that adding vlans using a range to an already existing list is not idempotent.  Logic should be added to expand a range if provided to match what is in the running config.

First I applied this playbook:

```yaml
    - name: Merge with existing vlans
      nxos_l2_interfaces:
        config:
          - name: "Ethernet1/3"
            trunk:
              native_vlan: 10
              allowed_vlans: 35-40 
            #access:
            #  vlan: 30
        state:  merged
```

Re-running this playbook is idempotent.

I then tried to add a few more vlans using the following playbook.

```yaml
    - name: Merge with existing vlans
      nxos_l2_interfaces:
        config:
          - name: "Ethernet1/3"
            trunk:
              native_vlan: 10
              allowed_vlans: 50-60 
            #access:
            #  vlan: 30
        state:  merged

```

It adds them but it's not idempotent.

```shell
    "after": [
        {
            "name": "Ethernet1/3", 
            "trunk": {
                "allowed_vlans": "35-40,50-60", 
                "native_vlan": 10
            }
        }
    ], 
    "ansible_facts": {
        "discovered_interpreter_python": "/usr/bin/python"
    }, 
    "before": [
        {
            "name": "Ethernet1/3", 
            "trunk": {
                "allowed_vlans": "35,36,37,38,39,40,50,51,52,53,54,55,56,57,58,59,60", 
                "native_vlan": 10
            }
        }
    ], 
    "changed": true, 
    "commands": [
        "interface Ethernet1/3", 
        "switchport trunk allowed vlan add 50,51,52,53,54,55,56,57,58,59,60"
    ], 
    "invocation": {
        "module_args": {
            "config": [
                {
                    "access": null, 
                    "name": "Ethernet1/3", 
                    "trunk": {
                        "allowed_vlans": "50,51,52,53,54,55,56,57,58,59,60", 
                        "native_vlan": 10
                    }
                }
            ], 
            "state": "merged"
        }
    }
}
```

If the state is merged we should only send commands to add the new vlans if they are not already part of the list.

Reply:
The `replaced` state does not appear to work properly.

I start with the following config:

```shell
interface Ethernet1/3
  switchport
  switchport trunk native vlan 40
  switchport trunk allowed vlan 35-40,50-60
```

I run the following playbook:

```yaml
    - name: Merge with existing vlans
      nxos_l2_interfaces:
        config:
          - name: "Ethernet1/3"
            trunk:
              native_vlan: 50
              allowed_vlans: 45-70
        state: replaced
```

Here is my config after running the playbook.

```shell
interface Ethernet1/3
  switchport
  switchport trunk native vlan 50
  switchport trunk allowed vlan 35-40,45-70
```

The `replaced` state should replace my running config with whatever is defined in the playbook but as you can see it left behind 35-40.

Reply:
**NOTE: ** Same problem with `state: overridden`

Start with the following configs:

```
interface Ethernet1/3
  switchport
  switchport trunk native vlan 50
  switchport trunk allowed vlan 35-40,45-70


interface Ethernet1/4
  switchport
  switchport trunk native vlan 60
  switchport trunk allowed vlan 35-40

```

Run the following playbook

```yaml
    - name: Merge with existing vlans
      nxos_l2_interfaces:
        config:
          - name: "Ethernet1/3"
            trunk:
              native_vlan: 50
              allowed_vlans: 45-70 
        state: overridden 
```

State after running the playbook:

```
interface Ethernet1/3
  switchport
  switchport trunk allowed vlan 35-40,45-70

interface Ethernet1/4
  switchport

```

State of Ethernet1/4 is correct but like state replaced Ethernet1/3 should only have vlans 45-70

Reply:
@mikewiebe : I have made few changes. I tested with various combinations. Can you please take a look? I will work on adding necessary testcases and documentation.